File: visual-app/src/main/java/org/visual/app/component/DraggableStackPane.java
package org.visual.app.component;

import javafx.scene.layout.StackPane;
import lombok.val;

public class DraggableStackPane extends StackPane {
  private final double[] dragAnchor = new double[2];

  public DraggableStackPane() {
    super();
    bindEvent();
  }

  private void bindEvent() {
    setOnMousePressed(event -> {
      dragAnchor[0] = event.getSceneX() - getLayoutX();
      dragAnchor[1] = event.getSceneY() - getLayoutY();
    });

    setOnMouseDragged(event -> {
      // 计算鼠标拖动的距离并更新 StackPane 的位置
      val newX = event.getSceneX() - dragAnchor[0];
      val newY = event.getSceneY() - dragAnchor[1];

      // 设置 StackPane 新的布局位置
      setLayoutX(newX);
      setLayoutY(newY);
    });
  }
}

File: visual-app/src/main/kotlin/org/visual/app/component/DraggableStackPane.kt
package org.visual.app.component

import io.github.oshai.kotlinlogging.KotlinLogging
import javafx.scene.Cursor
import javafx.scene.layout.StackPane

class DraggableStackPane : StackPane() {
  private val logger = KotlinLogging.logger {}
  private val dragAnchor = DoubleArray(2)

  init {
    setOnMouseEntered {
      cursor = Cursor.CROSSHAIR
    }
  }

  init {
    setOnMousePressed { event ->
      dragAnchor[0] = event.sceneX - layoutX
      dragAnchor[1] = event.sceneY - layoutY
    }

    setOnMouseDragged { event ->
      val newX = event.sceneX - dragAnchor[0]
      val newY = event.sceneY - dragAnchor[1]

      // 设置 StackPane 新的布局位置
      layoutX = newX
      layoutY = newY
    }
  }

}